File: java-union-reader-writer-example/src/main/java/com/intellicess/union/java_union_reader_writer_example/utils/CsvJwlfConverter.java
package com.intellicess.union.java_union_reader_writer_example.utils;

import com.intellicess.union.java_union_reader_writer_example.client.DateTimeConverter;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfCurve;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfHeader;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfRequest;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfValueType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class CsvJwlfConverter {

    private static final List<String> TRUE_BOOLEAN_VALUES = List.of("yes", "true", "on");
    private static final List<String> FALSE_BOOLEAN_VALUES = List.of("no", "false", "off");
    public static final String FILENAME_METADATA_KEY = "filename";

    private static final Logger log = LogManager.getLogger(CsvJwlfConverter.class);

    private CsvJwlfConverter() {
    }

    public static JwlfRequest convertCsvToJwlf(String well, String filename, String content,
                                               Map<String, JwlfValueType> headerValueTypeMap) {
        String filenameWithoutExtension = filename.replace(".csv", "");
        JwlfRequest jwlfRequest = new JwlfRequest();
        JwlfHeader jwlfHeader = JwlfHeader.builder()
                .name(filenameWithoutExtension)
                .well(well)
                .build();
        jwlfRequest.setHeader(jwlfHeader);
        if (content == null || content.isBlank()) {
            return jwlfRequest;
        }

        List<String> lines = Arrays.asList(content.split("\n"));
        String titleLine = lines.get(0);
        List<String> headers = Arrays.asList(titleLine.split(","));
        List<JwlfCurve> jwlfCurves = headers.stream()
                .map(header -> JwlfCurve.builder()
                        .name(header)
                        .valueType(headerValueTypeMap.get(header))
                        .build())
                .collect(Collectors.toList());
        jwlfRequest.setCurves(jwlfCurves);
        if (lines.size() == 1) {
            return jwlfRequest;
        }

        List<String> dataLines = lines.stream().skip(1).collect(Collectors.toList());
        List<List<Object>> data = dataLines.stream()
                .map(dataLine -> convertCsvLineToJwlfData(dataLine, jwlfCurves))
                .collect(Collectors.toList());
        jwlfRequest.setData(data);
        return jwlfRequest;
    }

    private static List<Object> convertCsvLineToJwlfData(String dataLine, List<JwlfCurve> jwlfCurves) {
        List<String> values = Arrays.asList(dataLine.split(","));
        if (values.size() != jwlfCurves.size()) {
            throw new RuntimeException("Data line='" + dataLine + "' should have the same number of values as number of headers ('" + jwlfCurves.size() + "').");
        }
        List<Object> castedValues = new ArrayList<>();
        for (int i = 0; i < values.size(); i++) {
            JwlfCurve jwlfCurve = jwlfCurves.get(i);
            Object value = values.get(i);
            Object castedValue = castValueToRightType(jwlfCurve, value);
            castedValues.add(castedValue);
        }
        return castedValues;
    }

    public static Object castValueToRightType(JwlfCurve jwlfCurve, Object value) {
        if (value == null) {
            return null;
        }

        JwlfValueType jwlfValueType = jwlfCurve.getValueType();
        if (jwlfValueType == null) {
            return value;
        }

        String stringValue = value.toString();
        switch (jwlfValueType) {
            case STRING:
                return stringValue;
            case INTEGER:
                return Integer.parseInt(stringValue);
            case FLOAT:
                return Double.parseDouble(stringValue);
            case BOOLEAN:
                String lowerCaseStringValue = stringValue.toLowerCase();
                if (TRUE_BOOLEAN_VALUES.contains(lowerCaseStringValue)) {
                    return true;
                } else if (FALSE_BOOLEAN_VALUES.contains(lowerCaseStringValue)) {
                    return false;
                }
                break;
            case DATETIME:
                return DateTimeConverter.parseZonedDateTime(stringValue);
        }
        log.warn("Couldn't cast value='{}' to value type='{}'. Returning null for curve with name='{}'",
                value, jwlfValueType, jwlfCurve.getName());
        return null;
    }
}

File: csharp-union-reader-writer-example/csharp-union-reader-writer-example/CsvJwlfConverter.cs
using Microsoft.VisualBasic;
using Newtonsoft.Json.Linq;
using Serilog;
using Serilog.Core;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using System.Reflection.PortableExecutable;
using System.Runtime.InteropServices;

namespace Union;

public static class CsvJwlfConverter
{

    private static List<string> TRUE_BOOLEAN_VALUES = new List<string> { "yes", "true", "on" };
    private static List<string> FALSE_BOOLEAN_VALUES = new List<string> { "no", "false", "off" };
    public static string FILENAME_METADATA_KEY = "filename";

    private static Logger log = AppLogFactory.CreateLogger();


    public static JwlfRequest ConvertCsvToJwlf(string well, string filename, string content,
                                               Dictionary<string, JwlfValueType> headerValueTypeMap)
    {
        string filenameWithoutExtension = filename.Replace(".csv", "");
        JwlfRequest jwlfRequest = new JwlfRequest();
        JwlfHeader jwlfHeader = new JwlfHeader
        {
            Name = filenameWithoutExtension,
            Well = well
        };
        jwlfRequest.Header = jwlfHeader;
        if (content == null || String.IsNullOrWhiteSpace(content))
        {
            return jwlfRequest;
        }

        List<string> lines = content.Split("\n").ToList();
        string titleLine = lines.First();
        List<string> headers = titleLine.Split(",").ToList();
        List<JwlfCurve> jwlfCurves = headers.Select(header => {
            JwlfValueType? valueType = null;
            if (headerValueTypeMap.ContainsKey(header))
            {
                valueType = headerValueTypeMap[header];
            }
            return new JwlfCurve
            {
                Name = header,
                ValueType = valueType
            };
        })
            .ToList();
        jwlfRequest.Curves = jwlfCurves;
        if (lines.Count() == 1)
        {
            return jwlfRequest;
        }

        List<string> dataLines = lines.Skip(1).ToList();
        List<List<Object>> data = dataLines.Select(dataLine => ConvertCsvLineToJwlfData(dataLine, jwlfCurves)).ToList();
        jwlfRequest.Data = data;
        return jwlfRequest;
    }

    private static List<Object> ConvertCsvLineToJwlfData(string dataLine, List<JwlfCurve> jwlfCurves)
    {
        List<string> values = dataLine.Split(",").ToList();
        if (values.Count() != jwlfCurves.Count())
        {
            throw new Exception("Data line='" + dataLine + "' should have the same number of values as number of headers ('" + jwlfCurves.Count() + "').");
        }
        List<Object> castedValues = new List<Object>();
        for (int i = 0; i < values.Count(); i++)
        {
            JwlfCurve jwlfCurve = jwlfCurves[i];
            Object value = values[i];
            Object castedValue = CastValueToRightType(jwlfCurve, value);
            castedValues.Add(castedValue);
        }
        return castedValues;
    }

    public static Object CastValueToRightType(JwlfCurve jwlfCurve, Object value)
    {
        if (value == null)
        {
            return null;
        }

        JwlfValueType? jwlfValueType = jwlfCurve.ValueType;
        if (!jwlfValueType.HasValue)
        {
            return value;
        }

        string stringValue = value.ToString();
        switch (jwlfValueType)
        {
            case JwlfValueType.STRING:
            case JwlfValueType.DATETIME:
                return stringValue;
            case JwlfValueType.INTEGER:
                return int.Parse(stringValue);
            case JwlfValueType.FLOAT:
                return double.Parse(stringValue, CultureInfo.GetCultureInfo("nl-NL"));
            case JwlfValueType.BOOLEAN:
                string lowerCasestringValue = stringValue.ToLowerInvariant();
                if (TRUE_BOOLEAN_VALUES.Contains(lowerCasestringValue))
                {
                    return true;
                }
                else if (FALSE_BOOLEAN_VALUES.Contains(lowerCasestringValue))
                {
                    return false;
                }
                break;
        }
        log.Warning("Couldn't cast value='{Value}' to value type='{JwlfValueType}'. Returning null for curve with name='{JwlfCurveName}'",
                value, jwlfValueType, jwlfCurve.Name);
        return null;
    }
}

